### PR TITLE
resolve #517, when a nested dependency is imported directly, re-link …

### LINKED ISSUE #517: solve the issue with direct import of a dependency
This is the main issue, this epic is trying to solve.

For example, a component is-string depends on another component is-type.
A user imported is-string, which also saved is-type as an NPM package. Later on, the user imports is-type directly ('bit import is-type') in order to update it. 

We have to make sure that 'is-string' links are pointing to the newly imported 'is-type' and not to the old npm package.

### PR DESCRIPTION
**Pull request by davidfirst**
…all its dependents

### CODE REVIEW

File: src/consumer/component/consumer-component.js [hunk 1]
```diff
@@ -98,6 +99,7 @@ export default class Component {
   writtenPath: ?string; // needed for generate links
   dependenciesSavedAsComponents: ?boolean = true; // otherwise they're saved as npm packages
   originallySharedDir: ?string; // needed to reduce a potentially long path that was used by the author
+  _wasOriginallySharedDirStripped: ?boolean;
```
Comment (GiladShoham) [inline src/consumer/component/consumer-component.js]: please add a comment about this property

File: e2e/commands/import.e2e.js [hunk 2]
```diff
@@ -1448,8 +1448,7 @@ describe('bit import', function () {
     });
   });
 
-  // @todo: fix this case. currently it does change it in the bit.map, but the link on the filesystem still direct to the .dependency dir
-  describe.skip('import component is-type as a dependency of is-string and then import is-type directly', () => {
+  describe('import component is-type as a dependency of is-string and then import is-type directly', () => {
```
Comment (GiladShoham) [inline e2e/commands/import.e2e.js]: We should add the opposite case:
import is-type then import is-string.
while we will import is-string we will have is-type in the package.json of course, so we will install it as a package. we need to make sure that after this install we link it to the imported is-type.

File: src/scope/scope.js [hunk 3]
```diff
@@ -1129,6 +1145,23 @@ export default class Scope {
     return updatedComponents;
   }
 
+  /**
+   * find the components in componentsPool which one of their dependencies include in potentialDependencies
+   */
```
Comment (GiladShoham) [inline src/scope/scope.js]: add the args / return to doc

File: src/scope/scope.js [hunk 4]
```diff
@@ -946,6 +947,21 @@ export default class Scope {
     return this.loadRemoteComponent(id);
   }
 
+  /**
+   * load components from the model and return them as ComponentVersion array.
+   * if a component is not available locally, it'll just ignore it without throwing any error.
+   */
```
Comment (GiladShoham) [inline src/scope/scope.js]: add the args / return to doc

File: src/consumer/consumer.js [hunk 5]
```diff
@@ -616,9 +616,27 @@ export default class Consumer {
...
       : writtenComponents;
     this.linkComponents(allComponents, bitMap);
+    await this.reLinkDirectlyImportedDependencies(writtenComponents, bitMap);
     return allComponents;
   }
 
+  /**
+   * an IMPORTED component might be NESTED before.
+   * find those components and re-link all their dependents
+   */
```
Comment (GiladShoham) [inline src/consumer/consumer.js]: add the args / return to doc

Comment (davidfirst) [reply src/consumer/consumer.js]: It's a decision we need to make.
We use Flow to document the args and return type. When there is nothing to document about the params except their type, re-documenting the type in the comment, feels to me as a duplication.
Every change in the args or return type needs to be done twice, in Flow and in the docs.

File: e2e/commands/import.e2e.js (hunk 2 above)
Comment (davidfirst) [reply e2e/commands/import.e2e.js]: Correct. We have another task for this case: https://github.com/teambit/bit/issues/521
I guess I'll add the e2e test while fixing it.

**Pull request merged**